src/newTest/java/org/opensearch/test/framework/cluster/PortAllocator.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2021 floragunn GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.test.framework.cluster;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+import org.opensearch.test.framework.cluster.SocketUtils.SocketType;
+
+
+public class PortAllocator {
+
+    public static final PortAllocator TCP = new PortAllocator(SocketType.TCP, Duration.ofSeconds(100));
+    public static final PortAllocator UDP = new PortAllocator(SocketType.UDP, Duration.ofSeconds(100));
+
+    private final SocketType socketType;
+    private final Duration timeoutDuration;
+    private final Map<Integer, AllocatedPort> allocatedPorts = new HashMap<>();
+
+    PortAllocator(SocketType socketType, Duration timeoutDuration) {
+        this.socketType = socketType;
+        this.timeoutDuration = timeoutDuration;
```

Comment:
What happens if a test takes longer than this duration to execute, its port will be attempted to be reused?

Reply:
The port *might* be retried. Please note that this is just a heuristic that is supposed to *help* avoiding port conflicts. It cannot avoid port conflicts completely, so any code that needs port might be prepared for conflicts and retry.

Reply:
If we want to avoid the potential overlap, it might be worthwhile to emit errors on tests that take more than the timeout duration.  Another mechanism would be to return a port leasing object that is closable so the test an signal the usage is complete.

Reply:
We need to be prepared for port conflicts anyway. So, this is really just a general measure to speed up the test by avoiding too many retries. So, we do not need 100% accuracy here,